server/sonar-server/src/main/java/org/sonar/server/rule/CachingRuleFinder.java

```diff
@@ -0,0 +1,196 @@
+import org.sonar.api.rules.RuleFinder;
+import org.sonar.api.rules.RulePriority;
+import org.sonar.api.rules.RuleQuery;
+import org.sonar.core.util.stream.MoreCollectors;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.rule.RuleDefinitionDto;
+import org.sonar.db.rule.RuleDto;
+import org.sonar.db.rule.RuleParamDto;
+import org.sonar.markdown.Markdown;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static org.sonar.core.util.stream.MoreCollectors.toSet;
+import static org.sonar.core.util.stream.MoreCollectors.uniqueIndex;
+
+/**
+ * A {@link RuleFinder} implementation that retrieves all rule definitions and their parameter when instantiated, cache
+ * them in memory and provide implementation of {@link RuleFinder}'s method which only read from this data in memory.
+ */
+public class CachingRuleFinder implements RuleFinder {
+
+  private static final Ordering<Map.Entry<RuleDefinitionDto, Rule>> FIND_BY_QUERY_ORDER = Ordering.natural().reverse().onResultOf(entry -> entry.getKey().getUpdatedAt());
+
+  private final Map<RuleDefinitionDto, Rule> rulesByRuleDefinition;
+  private final Map<Integer, Rule> rulesById;
+  private final Map<RuleKey, Rule> rulesByKey;
+
+  public CachingRuleFinder(DbClient dbClient) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      this.rulesByRuleDefinition = buildRulesByRuleDefinitionDto(dbClient, dbSession);
+      this.rulesById = this.rulesByRuleDefinition.entrySet().stream()
+        .collect(uniqueIndex(entry -> entry.getKey().getId(), Map.Entry::getValue));
+      this.rulesByKey = this.rulesByRuleDefinition.entrySet().stream()
+        .collect(uniqueIndex(entry -> entry.getKey().getKey(), Map.Entry::getValue));
+    }
+  }
+
+  private static Map<RuleDefinitionDto, Rule> buildRulesByRuleDefinitionDto(DbClient dbClient, DbSession dbSession) {
+    List<RuleDefinitionDto> dtos = dbClient.ruleDao().selectAllDefinitions(dbSession);
+    Set<RuleKey> ruleKeys = dtos.stream().map(RuleDefinitionDto::getKey).collect(toSet(dtos.size()));
+    ListMultimap<Integer, RuleParamDto> ruleParamsByRuleId = retrieveRuleParameters(dbClient, dbSession, ruleKeys);
+    Map<RuleDefinitionDto, Rule> rulesByDefinition = new HashMap<>(dtos.size());
+    for (RuleDefinitionDto definition : dtos) {
+      rulesByDefinition.put(definition, toRule(definition, ruleParamsByRuleId.get(definition.getId())));
+    }
+    return ImmutableMap.copyOf(rulesByDefinition);
+  }
+
+  private static ImmutableListMultimap<Integer, RuleParamDto> retrieveRuleParameters(DbClient dbClient, DbSession dbSession, Set<RuleKey> ruleKeys) {
+    if (ruleKeys.isEmpty()) {
+      return ImmutableListMultimap.of();
+    }
+    return dbClient.ruleDao().selectRuleParamsByRuleKeys(dbSession, ruleKeys)
+      .stream()
+      .collect(MoreCollectors.index(RuleParamDto::getRuleId));
+  }
+
+  @Override
+  @Deprecated
+  @CheckForNull
+  public Rule findById(int ruleId) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Add the missing @deprecated Javadoc tag. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Do not forget to remove this deprecated code someday. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)